source/opt/liveness.cpp

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) 2022 The Khronos Group Inc.
+// Copyright (c) 2022 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/opt/liveness.h"
+
+#include "source/opt/ir_context.h"
+
+namespace {
+
+const uint32_t kDecorationLocationInIdx = 2;
+const uint32_t kOpDecorateMemberMemberInIdx = 1;
+const uint32_t kOpDecorateMemberLocationInIdx = 3;
+const uint32_t kOpDecorateBuiltInLiteralInIdx = 2;
+const uint32_t kOpDecorateMemberBuiltInLiteralInIdx = 3;
+
+}  // namespace
+
+namespace spvtools {
+namespace opt {
+namespace analysis {
+
+LivenessManager::LivenessManager(IRContext* ctx) : ctx_(ctx), computed_(false) {
+  // Liveness sets computed when queried
+}
+
+void LivenessManager::InitializeAnalysis() {
+  live_locs_.clear();
+  live_builtins_.clear();
+  // Mark all builtins live for frag shader.
+  if (context()->GetStage() == SpvExecutionModelFragment) {
+    live_builtins_.insert(SpvBuiltInPointSize);
+    live_builtins_.insert(SpvBuiltInClipDistance);
+    live_builtins_.insert(SpvBuiltInCullDistance);
+  }
+}
+
+bool LivenessManager::IsAnalyzedBuiltin(uint32_t bi) {
```

Comment:
Essentially, by my reckoning, there are only three builtins that can be analyzed and removed between two stages: PointSize, ClipDistance and CullDistance. And only if Frag is not the downstream stage. That's because Frag fixed function consumes these values. Any other builtins cannot be removed because they are consumed by the downstream stage's fixed function.

ClipDistance stores in Vert feeding into Tesc can be removed if the Tesc shader does not explicitly read it. This is because Tesc fixed function does not consume ClipDistance.

Reply:
Oh ok, thanks for the explanation! Would you have a reference for that?

Reply:
> Would you have a reference for that?

This is based on my own analysis of the GLSL and Vulkan specs and my general understanding of the graphics pipeline.